개발자 일지
일단 과거부터 들어가자고
    STT는 딥그램이 무료토큰도 주고 좋을거라고 생각함 ㅇㅇ

    그뒤 ai는 차피 성능 안중요하고 걍 저지연 저비용만 중요함 그래서 나온게 젬마 -> 이건 걍 풀리기만 하고 없다네, 나중에 서버 구축하는 일 생기면 의미있을듯
    그래서 제미나이 라이트를 쓰기로 함
    3.5보단 2.5가 구번이니까 좋겠거니 싶음

    그리고 json으로 입력하는 방법 찾아서 이용하기로 함


    웹페이지 만드는건 걍 전체 쫚 클릭해서 html 받고 실제 사진 보여주니까 알아서 잘 하던데 긁적

    와 그리고 실제 서비스 시연에선 개쓸데 없는 부분이지만 로그인 로직 구현에 성공했다! (서버로 올리진 못함)

--------------------

7/6 ~ 7/9
디자이너 분들의 디자인을 무시하기 싫어짐
그래서 나중에 수정용이성도 높여보겠다 컴포넌트 화 한걸 그대로 옮기자! (란 생각을 하지 말아야했던게 아닐까)

다지아니 색 css부터 통일시키고
미친 노가다를 통해서
결국 완성했습니다
근데 나 결과리포트 쓰기 싫어 엉엉

와 근데 이거 하니까 백엔드 할 시간이 아에 없어졌는데요 ;;;;

그리고 지금까지 만든 백엔드 코드로 시연이라도 해볼려했는데
한도에 걸린것같다네요? 에라이
다행히 보라님 api키 받아서 진행하기로 함
(대충 나나 굿 콘)

---------------------

7/9 ~ 7/12 (이때까지 반드시 완성해야만 한다)

일단 gpt로 바꾸면서 다시 배우는중
프롬포트 캐싱이란 놈이 있는데 앞부분 천이상 부분 헤싱해서 일치하면 그 모델 그대로 이용한다네요?
근데 우리 모델이 10분동안 하면 총 천번하거든요?
그럼 무조건 헤싱하는게 이득임 헤싱하면 토큰이 쌀먹이 된다고!!

그래서 일단 토큰양을 측정해보고자 합니다.

{
    system_prom : 
    "audiences_persona": [
            {
                "id": 1,
                "persona": {
                    "interest":"AI",
                    "patience":"high"
                }
            }
        ]
    last_output : {
        "presentationSummary": "...",
        "audiences_state": [
            {
                "id": 1,
                "state":{
                    "E":0.3,
                    "V":0.2,
                    "C":0.1
                },
                "currentThought":"아직은 평범하다.",
                "overallImpression":"무난한 발표."
            }
        ]
    }
    "latestSpeech": "...",
}

이렇게 처음에 생각했었는데
프롬포트캐싱이 가장 긴 공통뭐시기로 잡는다고 해서
서머리 같은 경우 모두가 사용하면서 기존길이가 보존되도록 만들어봄직하니까
프롬흐로 서머리는 기존 내용 뒤에 붙이기만하자로 해서
서머리 일부가 공통으로 인식되도록 하는 방안을 생각해봄

세션 정보 json
template_json
{
    "id" : 1,
    "owner_id": "lhj0315",
    "created_at": "2026-07-11T21:30:00Z",
    "description" : "",
   

    "basic_info": {
    "presentation_title": "VR발표환경에서의 반응적 중립성 기반 AI 청중 백채널 디자인",
    "presentation_purpose": "연구결과공유",
    "used_language": "한국어",
    "duration_minutes": 10,
    "environment_type": "세미나실",
    "qa_count": 3
    },

    "file_info": {
    "slide_pdf_path": "~~~/VR발표환경에서의 반응적 중립성 기반 AI 청중 백채널 디자인.pdf",
    "paper_pdf_path": "~~~/AI_Audience_Backchannel_Research_Paper.pdf",
    "presentation_script_content": "안녕하세요. 발표자 Hazel입니다. 지금부터 VR 발표 환경에서의 반응적 중립성 기반 AI 청중 백채널 디자인에 대한 연구 결과를 공유해 드리겠습니다. 본 연구는 사용자가 VR 환경에서 발표할 때..."
    },

    "audience_info": {
    "audience_type": "심사위원 중심",
    "audience_scale_count": 6,
    "expertise_level": "중간",
    "interest_level": "중간"
    }
}

session_json
{
    "id" : ,
    "user_id": "lhj0315",

    "created_at": "2026-07-11T21:30:00Z",

    "template" : template_json

    "feedback" : feedback_json

}

pindb_json
{
    "id" : ,

    "user_id" : ,

    "template": template_json,

    "pincode" : ,

    "expires_at": "2026-07-11T21:30:00Z",

}

 #레디가 되면서 핀코드를 새로 만들고 크리에이트 시간을 업데이트함. 
        이때 크리에이트 1시간 넘은거 다 삭제? (이걸 넣어야할지, 넣는다면 언제 넣을지 고민중)
        유니티 핀코드와 db 핀코드 일치 & 만든지 1시간 안지남 -> 탬플릿 정보 유니티로 넘어가고 종료 (핀 = none으로 변경)
        유니티는 템플릿 바탕으로 환경 설정 -> 시작버튼 누르면 템플릿 json은 evc start로 넘어감
        발표가 끝나고 feedback 생기면 그 정보를 더해서 session_json 형태로 저장

피그마 문서를 좀 더 읽어보니까
걍 페르소나 없애고 청중 자체를 하나의 EVC로 묶는게 맞아 보였음
그 EVC도 M_t, D_t로 만들고
그럼 M-T랑 D_t를 내보내면 되는거고

일단 AI가 Mt를 잘 짜게 프롬을 구성해야함
    이전 단계의 패착
    서머리 업데이트 : 애초에 제대로 업데이트 못하고, 프롬 잘 이해못하고 반복했음
    한정적이 값이 금방 이전 상태를 참고하여 좋은거니까 하고 좋겠거니 하며 상한을 치거나 유지함
D_t는 현재 하는 말 + 맥락? + 발표자료 : 유니티 제공, 말의 절음? 이런건 딥그램이 잘 제공해줄거임
    유니티가 걍 시간차이보고 절었는지 이런거 잘 판단할 수 있는지
    유니티랑 자료같은건 시작할때 슬라이드 별 요약을 제공 받고
    이후에 슬라이브 펼쳐지면 슬라이드 정보주면 될듯..?

그리고 5초간격이 굉장히 짧다는 것을 알게됨 10초씩이나 아니면 10초인데 앞부분 5~10초정도 포함?

Mt Dt 가지고 EVC 만드는 공식 다시 한번 더 정리 필요. 잘할 수있는지
EVC가지고 행동정하는 식도 정리 필요

In [ ]:
페이지
→ 폭과 좌우 배치를 담당한다.

SurfaceCard
→ 흰 배경, radius, shadow, padding, 같은 높이를 담당한다.

각 카드 컴포넌트
→ 내부 section, gap, grid만 담당한다.

UploadSection / SegmentItem / Slider
→ 자기 내부 모양만 담당하고 부모 폭을 따른다.

In [ ]:
config = {
    "audience" : {
        "type" : "혼합",
        "number" : 50,
        "Expertise" : "high",
        "Interest" : "middle"

    },
    "environment" : {
        "name" : "AI 시대에 커뮤니케이션 전략",
        "place" : "세미나실",
        "purppose" : "연구결과 공유"

    },
    "question" : 3,
    "duration" : 10*60,
}

feedback_1 = {
  "evc": {
    "E": 72,
    "V": 68,
    "C": 81
  },

  "total_score": 78,

  "radar_score": {
    "E": 72,
    "V": 68,
    "C": 81,
  },

  "detail_scores": {
    "content": 82,
    "delivery": 76,
    "pronunciation" : 68,
    "time_management": 84
    },

  "analysis": {
    "content": {
      "메시지 명확성" : "우수",
      "발표 구조와 흐름": "보통",
      "근거와 자료 활용": "우수",
      "주장-근거 연결성": "우수",
      "어휘 및 표현 적절성" : "개선"
      },

    "delivery":{
      "시선 처리" : "우수",
      "발화 속도" : "보통",
      "발음 정확도" : "우수",
      "습관어 사용" : "보통",
      "시간 운영" : "우수"
      },
    },
  "coaching_timeline": [
    {
      "time": "05:12",
      "type": "good",
      "title": "문제의식 제시가 인상적이었어요!",
      "content": "도입부에서 문제 상황을 자연스럽게 제시하여 청중의 관심을 끌었습니다."
    },
    {
      "time": "14:58",
      "type": "good",
      "title": "핵심 내용을 명확하게 전달했어요.",
      "content": "핵심 메시지가 한 문장으로 정리되어 전달력이 높았습니다."
    },
    {
      "time": "21:43",
      "type": "warning",
      "title": "조금 더 천천히 말해볼까요?",
      "content": "발화 속도가 빨라 일부 내용이 전달되지 않았습니다."
    },
    {
      "time": "27:05",
      "type": "warning",
      "title": "습관어를 줄여보세요.",
      "content": "\"음\", \"어\"가 반복적으로 사용되었습니다."
    }
  ],

  "audience_graph": {
    "time": [0, 5, 10, 15, 20, 25, 30],

    "eye_contact": [65, 72, 75, 70, 82, 85, 74],

    "question_interest": [68, 69, 66, 74, 83, 86, 87],

    "positive_reaction": [60, 52, 57, 55, 64, 70, 68]
  },

  "graph_comment": [
    {
      "icon": "eye",
      "title": "시선이 집중되는 발표였어요",
      "content": "핵심 구간에서 청중 응시가 길게 유지되었습니다."
    },
    {
      "icon": "question",
      "title": "질문이 생기는 흥미로운 흐름이었어요",
      "content": "결론에서 질문 의도가 많이 나타났습니다."
    },
    {
      "icon": "smile",
      "title": "공감으로 이어진 전달력이었어요",
      "content": "사례와 근거가 연결될 때 긍정 반응이 증가했습니다."
    }
  ],

  "ai_insight": {
    "title": "기본기가 탄탄한 발표입니다.",
    "highlight": "한두 가지 디테일만 조금 더 다듬으면 훨씬 좋아질 거예요! 👏",
    "content": "다음 발표에서는 전문 용어를 조금 더 쉬운 표현으로 바꾸고 결론에서 핵심 메시지를 다시 한번 강조해 보세요."
  },

  "recommended_practice": {
    "title": "맞춤 훈련 시작",
    "practice_type": "발화 속도 향상"
  }
}

pratice : 요놈 이름에 따라서 정보가 다 기본으로 저장되어있을것
{
    이름 : 
    설명 : 
    훈련키워드 : 
    훈련키워드색 : 
    난이도 : 
    난이도색 : 
}

[rehear_users]
auth_id
id : user_id
config
{
    "audience": {...},
    "environment": {...},
    "question": 3,
    "duration": 15
}

나의 평가요소들
{
    강점 : (그런거 4개)
    평균 점수
    EVC 그래프
    연습 : practice_list : 배열(json?)
}



[templates]
id
owner_id
name
type (발표 / 면접)
description
keyword : 배열(json?)
config
    {
        "audience": {...},
        "environment": {...},
        "question": 3,
        "duration": 15
    }

템플릿 공유시 쌍방DB도 만들듯?
유저id <-> xrealid
지금은 그냥 주인id로만 끌고오기


[sessions]
id
user_id
name
created_at
total_score
tamplete_json
{
    그냥 탬플릿을 그대로 빼다박는게 더 효율적일듯? json으로 내보낼 수 있으니까
}
filepath_json
{
    이름 : 경로,
}
feedback_json
{
    EVC 점수
    세부항복별 점수
    점수 그래프
    AI 인사이트
    코칭 피드백
}


PRAGMA foreign_keys = ON;

--------------------------------------------------------
-- Re:hear 사용자
--------------------------------------------------------
CREATE TABLE rehear_users (
    id INTEGER PRIMARY KEY AUTOINCREMENT,

    auth_id INTEGER NOT NULL UNIQUE,

    config_json TEXT NOT NULL DEFAULT '{}',

    stats_json TEXT NOT NULL DEFAULT '{}',

    created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
    updated_at DATETIME DEFAULT CURRENT_TIMESTAMP
);

--------------------------------------------------------
-- 템플릿
--------------------------------------------------------
CREATE TABLE templates (
    id INTEGER PRIMARY KEY AUTOINCREMENT,

    owner_id INTEGER NOT NULL,

    name TEXT NOT NULL,

    type TEXT NOT NULL
        CHECK(type IN ('presentation', 'interview')),

    description TEXT DEFAULT '',

    keywords_json TEXT NOT NULL DEFAULT '[]',

    config_json TEXT NOT NULL,

    created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
    updated_at DATETIME DEFAULT CURRENT_TIMESTAMP,

    FOREIGN KEY(owner_id)
        REFERENCES rehear_users(id)
        ON DELETE CASCADE
);

--------------------------------------------------------
-- 발표 세션
--------------------------------------------------------
CREATE TABLE sessions (
    id INTEGER PRIMARY KEY AUTOINCREMENT,

    user_id INTEGER NOT NULL,

    template_id INTEGER,

    name TEXT NOT NULL,

    status TEXT NOT NULL DEFAULT 'completed'
        CHECK(status IN ('created', 'running', 'completed', 'failed')),

    created_at DATETIME DEFAULT CURRENT_TIMESTAMP,

    total_score REAL DEFAULT 0,

    template_json TEXT NOT NULL,

    filepath_json TEXT NOT NULL DEFAULT '{}',

    feedback_json TEXT NOT NULL DEFAULT '{}',

    FOREIGN KEY(user_id)
        REFERENCES rehear_users(id)
        ON DELETE CASCADE,

    FOREIGN KEY(template_id)
        REFERENCES templates(id)
        ON DELETE SET NULL
);

In [1]:
from google import genai
from google.genai import types

from pydantic import BaseModel, Field
from typing import List

from dotenv import load_dotenv
import os

load_dotenv()

client = genai.Client(
    api_key=os.getenv("GEMINI_API_KEY")
)


class AudienceState(BaseModel):
    E: float = Field(description="-1~1, Engagement")
    V: float = Field(description="-1~1, Evaluative Valence")
    C: float = Field(description="-1~1, Cognitive Clarity")


class AudienceOutput(BaseModel):
    id: int
    state: AudienceState
    currentThought: str
    overallImpression: str


class PresentationOutput(BaseModel):
    presentationSummary: str
    audiences: List[AudienceOutput]


SYSTEM_PROMPT = """
당신은 발표를 듣는 AI 청중 시뮬레이터이다.

입력으로 다음 정보가 제공된다.
- presentationSummary: 지금까지 발표 내용을 누적 요약한 것
- latestSpeech: 발표자가 방금 말한 내용
- audiences: 각 청중의 성격(persona), 현재 상태(state), 현재 생각(currentThought), 발표자에 대한 전체 인상(overallImpression)

당신의 역할은 latestSpeech를 반영하여 각 청중의 내부 상태를 갱신하는 것이다.

평가 기준

E (Engagement)
-1 : 발표에 거의 관심이 없고 집중하지 않음
 0 : 보통 수준
+1 : 매우 집중하고 있음

V (Evaluative Valence)
-1 : 발표를 매우 부정적으로 평가
 0 : 중립
+1 : 매우 긍정적으로 평가

C (Cognitive Clarity)
-1 : 발표를 거의 이해하지 못함
 0 : 부분적으로 이해
+1 : 매우 명확하게 이해

갱신 규칙

- persona를 반영한다.
- 기존 state를 기반으로 자연스럽게 변화시킨다.
- latestSpeech만큼만 변화시킨다.
- 발표가 명확하면 C를 증가시킨다.
- 발표가 흥미롭거나 자신의 관심사이면 E를 증가시킨다.
- 발표 품질이 좋거나 설득력이 높으면 V를 증가시킨다.
- currentThought는 현재 머릿속에서 드는 생각을 한 문장으로 작성한다.
- overallImpression은 발표가 끝났을 때 남는 누적 인상이므로 급격하게 바뀌지 않는다.
- presentationSummary는 latestSpeech를 반영하여 기존 summary를 자연스럽게 갱신한다.
- state의 모든 값은 반드시 [-1, 1] 범위 안에 있어야 한다.

출력은 반드시 JSON만 반환한다.
"""


input_json = """
{
    "presentationSummary": "...",
    "latestSpeech": "...",
    "audiences": [
        {
            "id": 1,
            "persona": {
                "interest":"AI",
                "patience":"high"
            },
            "state":{
                "E":0.3,
                "V":0.2,
                "C":0.1
            },
            "currentThought":"아직은 평범하다.",
            "overallImpression":"무난한 발표."
        }
    ]
}
"""

input_json = """
{
    "presentationSummary": "발표자는 XR 환경에서 발표 연습을 할 때 실제 청중처럼 반응하는 AI 청중 시스템의 필요성을 설명했다. 지금까지 시스템의 목표와 기본 구조를 소개했고, 발표 음성을 분석해 청중 상태를 갱신한다는 흐름을 설명하는 중이다.",
    "latestSpeech": "이 시스템에서는 발표자의 말을 5초 단위로 받아서 텍스트로 변환합니다. 그리고 그 텍스트를 기반으로 각 청중이 얼마나 집중하고 있는지, 발표를 긍정적으로 보는지, 내용을 잘 이해하고 있는지를 계속 업데이트합니다. 예를 들어 설명이 명확하면 이해도가 올라가고, 같은 말이 반복되거나 근거가 부족하면 집중도나 평가가 조금씩 떨어질 수 있습니다.",
    "audiences": [
        {
            "id": 1,
            "persona": {
                "interest": "AI",
                "priorKnowledge": "high",
                "patience": "high",
                "criticalBias": 0.35
            },
            "state": {
                "E": 0.42,
                "V": 0.18,
                "C": 0.51
            },
            "currentThought": "AI 청중이라는 아이디어는 흥미롭고 구조도 어느 정도 이해된다.",
            "overallImpression": "기술적 방향성이 분명하고 꽤 설득력 있는 발표다."
        },
        {
            "id": 2,
            "persona": {
                "interest": "presentation_training",
                "priorKnowledge": "medium",
                "patience": "medium",
                "criticalBias": 0.55
            },
            "state": {
                "E": 0.31,
                "V": 0.05,
                "C": 0.22
            },
            "currentThought": "발표 훈련에 도움이 될 것 같지만 실제로 어떻게 구현되는지는 아직 조금 궁금하다.",
            "overallImpression": "아이디어는 괜찮지만 구체적인 검증 방법이 더 필요해 보인다."
        },
        {
            "id": 3,
            "persona": {
                "interest": "design",
                "priorKnowledge": "low",
                "patience": "medium",
                "criticalBias": 0.25
            },
            "state": {
                "E": 0.12,
                "V": 0.02,
                "C": -0.18
            },
            "currentThought": "무슨 시스템인지는 알겠지만 기술 설명이 조금 어렵게 느껴진다.",
            "overallImpression": "흥미로운 발표지만 아직은 조금 추상적으로 들린다."
        }
    ]
}
"""


response = client.models.generate_content(
    model="gemini-2.5-flash-lite",
    contents=input_json,
    config=types.GenerateContentConfig(
        system_instruction=SYSTEM_PROMPT,
        response_mime_type="application/json",
        response_schema=PresentationOutput,
        temperature=0.3,
        max_output_tokens=2048,
    ),
)

print(response.text)

result = PresentationOutput.model_validate_json(response.text)

print(result)

{
  "presentationSummary": "발표자는 XR 환경에서 발표 연습을 할 때 실제 청중처럼 반응하는 AI 청중 시스템의 필요성을 설명했다. 지금까지 시스템의 목표와 기본 구조를 소개했고, 발표 음성을 분석해 청중 상태를 갱신한다는 흐름을 설명하는 중이다. 발표자는 발표자의 말을 5초 단위로 받아 텍스트로 변환하고, 이를 기반으로 각 청중의 집중도, 긍정적 평가, 이해도를 업데이트한다고 설명했다. 또한, 설명의 명확성이 이해도에 영향을 미치고, 반복이나 근거 부족은 집중도와 평가를 낮출 수 있다고 덧붙였다.",
  "audiences": [
    {
      "id": 1,
      "state": {
        "E": 0.5,
        "V": 0.25,
        "C": 0.6
      },
      "currentThought": "5초 단위로 음성을 분석하고 상태를 업데이트하는 방식이 구체적이어서 이해하기 쉽다.",
      "overallImpression": "기술적 방향성이 분명하고 꽤 설득력 있는 발표다."
    },
    {
      "id": 2,
      "state": {
        "E": 0.4,
        "V": 0.1,
        "C": 0.35
      },
      "currentThought": "음성 분석을 통해 청중 상태를 실시간으로 업데이트하는 방식이 흥미롭다. 실제 구현이 기대된다.",
      "overallImpression": "아이디어는 괜찮지만 구체적인 검증 방법이 더 필요해 보인다."
    },
    {
      "id": 3,
      "state": {
        "E": 0.2,
        "V": 0.05,
        "C": 0.0
      },
      "currentThought": "5초 단위 분석이라는 구체적인 설명 덕분에 시스템이 어떻게 작동하는지 조금 더 명확하게 이해하기 시작했다.",
      "o

그러면 입력
{
    "presentationSummary": "...",
    "latestSpeech": "...",
    "audiences" : 
    [
        {
            "id" : int,
            "persona": { ... },
            "state": {
                "E": -1~1,
                "V": -1~1,
                "C": -1~1
            },
            "currentThought": "...",
            "overallImpression": "...",
        } ,
    ]
}

출력
{
    "presentationSummary": "...",
    "audiences" : 
    [
        {
            "id" : int
            "state": {
                "E": -1~1,
                "V": -1~1,
                "C": -1~1
            },
            "currentThought": "...",
            "overallImpression": "...",
        } ,
    ]
}

In [6]:
from google import genai

client = genai.Client(api_key=os.getenv("GEMINI_API_KEY"))

for model in client.models.list():
    print(model.name)

models/gemini-2.5-flash
models/gemini-2.5-pro
models/gemini-2.0-flash
models/gemini-2.0-flash-001
models/gemini-2.0-flash-lite-001
models/gemini-2.0-flash-lite
models/gemini-2.5-flash-preview-tts
models/gemini-2.5-pro-preview-tts
models/gemma-4-26b-a4b-it
models/gemma-4-31b-it
models/gemini-flash-latest
models/gemini-flash-lite-latest
models/gemini-pro-latest
models/gemini-2.5-flash-lite
models/gemini-2.5-flash-image
models/gemini-3-pro-preview
models/gemini-3-flash-preview
models/gemini-3.1-pro-preview
models/gemini-3.1-pro-preview-customtools
models/gemini-3.1-flash-lite-preview
models/gemini-3.1-flash-lite
models/gemini-3-pro-image-preview
models/gemini-3-pro-image
models/nano-banana-pro-preview
models/gemini-3.1-flash-image-preview
models/gemini-3.1-flash-image
models/gemini-3.1-flash-lite-image
models/gemini-3.5-flash
models/gemini-omni-flash-preview
models/lyria-3-clip-preview
models/lyria-3-pro-preview
models/gemini-3.1-flash-tts-preview
models/gemini-robotics-er-1.5-preview
mod

In [ ]:
from google import genai
from google.genai import types

from pydantic import BaseModel, Field
from dotenv import load_dotenv
import os

load_dotenv()

client = genai.Client(
    api_key=os.getenv("GEMINI_API_KEY")
)

class AudienceState(BaseModel):
    Engagement: float = Field(description="-1~1, 청중이 발표 상황과 발표자에게 주의를 기울이며 청취를 지속하려는 정도 / 무관심, 이탈 <-> 집중, 몰임")
    Evaluative_Valence: float = Field(description="-1~1, 청중이 발표 내용 또는 발표 수행을 긍정적 또는 부정적으로 평가하는 방향성 / 부정적 평가 <-> 긍정적 평가")
    Cognitive_Clarity: float = Field(description="-1~1, 청중이 발표 내용의 구조와 의미를 명확히 파악하는 정도 / 혼란, 불명확 <-> 이해, 명료")
    reaction: str = Field(description="nod, look_away, confused 등")


response = client.models.generate_content(
    model="gemini-2.5-flash-lite",
    contents="""
발표를 듣고 청중 상태를 평가하세요.

발표:
안녕하세요. 오늘은 XR 발표 훈련 서비스 Re:hear를 소개하겠습니다.
저희 서비스는 AI 청중이 발표 내용을 이해한 정도에 따라 반응합니다.

JSON만 출력하세요.
""",
    config=types.GenerateContentConfig(
        response_mime_type="application/json",
        response_schema=AudienceState,
        temperature=0.2,
        max_output_tokens=100,
    ),
)

print(response.text)

state = AudienceState.model_validate_json(response.text)

print(state)
print(state.attention)
print(state.reaction)

{
  "attention": 0.8,
  "understanding": 0.7,
  "confusion": 0.1,
  "reaction": "nod"
}
attention=0.8 understanding=0.7 confusion=0.1 reaction='nod'
0.8
nod


In [12]:
interaction.model_dump()

{'status': 'completed',
 'model': 'gemini-2.5-flash-lite',
 'id': 'v1_ChdfYjVJYXZqRUhMaWYwLWtQekpQZmdBYxIXX2I1SWF2akVITGlmMC1rUHpKUGZnQWM',
 'created': '2026-07-04T08:06:21Z',
 'updated': '2026-07-04T08:06:21Z',
 'usage': {'total_input_tokens': 8,
  'input_tokens_by_modality': [{'modality': 'text', 'tokens': 8}],
  'total_cached_tokens': 0,
  'total_output_tokens': 949,
  'total_tool_use_tokens': 0,
  'total_thought_tokens': 0,
  'total_tokens': 957},
 'service_tier': 'standard',
 'steps': [{'type': 'thought',
   'signature': 'CiRlMjQ4MzBhNy01Y2Q2LTQyZmUtOTk4Yi1lZTUzOWU3MmI5YzM='},
  {'type': 'model_output',
   'content': [{'text': "## Classic & Moist Banana Bread\n\nThis recipe yields a delicious, moist, and flavorful banana bread that's perfect for breakfast, a snack, or dessert.\n\n**Yields:** 1 loaf (about 10-12 slices)\n**Prep time:** 15 minutes\n**Bake time:** 55-65 minutes\n\n**Ingredients:**\n\n*   **For the Bananas:**\n    *   3 ripe to overripe bananas (the spottier, the bett

In [9]:
print(interaction.output_text)

## Classic & Moist Banana Bread

This recipe yields a delicious, moist, and flavorful banana bread that's perfect for breakfast, a snack, or dessert.

**Yields:** 1 loaf (about 10-12 slices)
**Prep time:** 15 minutes
**Bake time:** 55-65 minutes

**Ingredients:**

*   **For the Bananas:**
    *   3 ripe to overripe bananas (the spottier, the better!)
*   **For the Wet Ingredients:**
    *   1/3 cup unsalted butter, melted
    *   1 large egg, lightly beaten
    *   1 teaspoon vanilla extract
    *   1/4 cup milk (any kind: whole, 2%, almond, soy, etc.)
*   **For the Dry Ingredients:**
    *   1 1/2 cups all-purpose flour
    *   3/4 cup granulated sugar (you can adjust slightly based on banana sweetness)
    *   1 teaspoon baking soda
    *   1/2 teaspoon ground cinnamon (optional, but highly recommended)
    *   1/4 teaspoon salt
*   **Optional Add-ins:**
    *   1/2 cup chopped nuts (walnuts or pecans are classic)
    *   1/2 cup chocolate chips (semi-sweet or dark)

**Equipment:**



뭘해야할까
일단 확인해봐야할게

시작 누르면 유니티 : 오디오 수집 -> 보내기 이거 만들어야하는데 어케만들지
또 해야할게
오디오 보내면, 텍스트로 바꾸기 : 이건 될듯?
텍스트 -> ai 상태변환
    상태변화 어케하지
    그냥 지금 상황주고 바꿔서 달라고 할까 -> 이거 안비싸나..?

In [ ]:
import json
from faster_whisper import WhisperModel
from openai import OpenAI
import os
import logging
from dotenv import load_dotenv

from deepgram import (
    DeepgramClient,
)

load_dotenv()

# Path to the audio file
language = "ko-KR"
AUDIO_FILE = "sample.m4a"
API_KEY = os.getenv("DEEPGRAM_API_KEY")



# STEP 1 Create a Deepgram client using the API key
deepgram = DeepgramClient(api_key=API_KEY)

with open(AUDIO_FILE, "rb") as file:
    buffer_data = file.read()

# STEP 2: Call the transcribe_file method with the audio file and options
response = deepgram.listen.v1.media.transcribe_file(
    request=buffer_data,
    model="nova-3",
    filler_words=True, #어.. 같은거
    utterances=True, # 발화단위로 나누기
    language=language,

    #텍스트 감정 sentiment=True,
    # 보기 좋게 smart_formatting=True,
    # 숫자를 숫자로 numerals=True,
).model_dump()

response['results']['channels'][0]['alternatives'][0]['transcript']



In [16]:
res1_0 = response.model_dump()
tc = res1_0['results']['channels'][0]['alternatives'][0]['transcript']
자세히 = res1_0['results']['channels'][0]['alternatives'][0]['words']

print(tc)
자세히

안녕하세요 그 발표를 시작하겠습니다


[{'word': '안녕하세요', 'start': 0.64, 'end': 1.5999999, 'confidence': 0.5161133},
 {'word': '그', 'start': 2.0, 'end': 2.6399999, 'confidence': 0.9301758},
 {'word': '발표를', 'start': 3.36, 'end': 4.4, 'confidence': 0.9902344},
 {'word': '시작하겠습니다', 'start': 4.48, 'end': 5.6, 'confidence': 0.99609375}]

| 파라미터               | 설명                 | SDK 기본         | 추천                 |
| ------------------ | ------------------ | -------------- | ------------------ |
| `request`          | 전사할 오디오(bytes)     | 필수             | WAV 바이트            |
| `model`            | 사용할 STT 모델         | 없음(필수에 가까움)    | `nova-3`           |
| `language`         | 언어                 | 자동 감지 또는 모델 기본 | `ko-KR`            |
| `filler_words`     | "어", "음" 등을 전사     | None(미지정)      | **True 테스트**       |
| `punctuate`        | 쉼표, 마침표 자동 추가      | None           | **False 추천**       |
| `smart_format`     | 숫자, 날짜 등을 보기 좋게 변환 | None           | **False 추천**       |
| `utterances`       | 발화 단위로 나누기         | None           | True               |
| `utt_split`        | 몇 초 침묵 시 발화 분리     | None           | 0.8~1.5 정도 실험      |
| `numerals`         | "삼백" → "300" 변환    | None           | False              |
| `paragraphs`       | 문단으로 묶기            | None           | False              |
| `profanity_filter` | 욕설 순화              | None           | False              |
| `diarize_model`    | 화자 구분              | None           | 발표에서는 보통 사용 안 함    |
| `detect_language`  | 언어 자동 감지           | None           | False (`ko-KR` 지정) |


이건 거의 모든 옵션이 적혀 있어서 처음 보면 부담스러워. **네 프로젝트에서 중요한 것**에는 ⭐를 붙여둘게.

---

### 함수

`transcribe_file(...)`

Deepgram의 STT REST API를 이용해 **오디오 또는 비디오 파일을 텍스트로 변환**한다.

---

### 매개변수(Parameters)

⭐ **request**

* 전사할 오디오 데이터.
* `bytes`(파일 바이트) 또는 바이트 스트림을 전달한다.

⭐ **callback**

* 전사가 끝난 후 결과를 받을 URL.
* 긴 작업을 비동기로 처리할 때 사용.

**callback_method**

* callback 요청에 사용할 HTTP 메서드(GET, POST 등).

**extra**

* 응답에 함께 넣어둘 임의의 데이터(Key-Value).

**sentiment**

* 텍스트의 감정(긍정/부정 등)을 분석.

**summarize**

* 전사 내용을 요약.

**tag**

* 요청을 구분하기 위한 태그.

**topics**

* 대화의 주제를 자동으로 추출.

**custom_topic**

* 사용자가 지정한 주제를 우선적으로 탐지.

**custom_topic_mode**

* 사용자 지정 주제만 탐지할지(`strict`), 일반 주제도 함께 탐지할지(`extended`) 설정.

**intents**

* 화자의 의도(Intent)를 분석.

**custom_intent**

* 사용자가 지정한 의도를 우선 탐지.

**custom_intent_mode**

* 의도 탐지 방식을 설정.

**detect_entities**

* 사람 이름, 회사명, 장소 등 개체(Entity)를 추출.

**detect_language**

* 음성의 언어를 자동으로 감지.

**diarize** *(구버전)*

* 화자 분리(누가 말했는지 구분).

**diarize_model**

* 사용할 화자 분리 모델 선택.

**dictation**

* 받아쓰기 모드(구두점 등을 음성으로 말하는 경우에 적합).

**encoding**

* 입력 오디오의 인코딩 형식 지정(PCM, FLAC 등).

⭐ **filler_words**

* `"어"`, `"음"`, `"uh"`, `"um"` 같은 **필러(filler word)를 그대로 전사**할지 여부.
* 네 프로젝트에서는 매우 중요한 옵션.

**keyterm**

* Nova-3 전용.
* 특정 전문 용어나 브랜드명을 더 잘 인식하도록 힌트를 제공.

**keywords**

* 특정 단어나 용어의 인식률을 높이거나 낮춤.

⭐ **language**

* 음성 언어(BCP-47 형식).
* 예: `ko-KR`, `en-US`.

**measurements**

* "3미터" 같은 측정 단위를 약어로 변환.

⭐ **model**

* 사용할 STT 모델.
* 예: `nova-3`.

**multichannel**

* 여러 오디오 채널(좌/우, 마이크 여러 개)을 각각 독립적으로 전사.

**numerals**

* `"삼백"` → `"300"`처럼 숫자를 숫자 형태로 변환.

**paragraphs**

* 결과를 문단 단위로 나눔.

**profanity_filter**

* 욕설을 순화하거나 제거.

⭐ **punctuate**

* 마침표, 쉼표, 대문자 등을 자동으로 추가.

⭐ **redact**

* 개인정보(전화번호, 주민번호 등)를 가림 처리.

**replace**

* 특정 단어나 문장을 다른 표현으로 치환.

**search**

* 특정 단어나 문장을 검색.

⭐ **smart_format**

* 결과를 사람이 읽기 좋게 자동 포맷팅.
* 날짜, 숫자, 전화번호 등을 보기 좋게 변환.

⭐ **utterances**

* 긴 음성을 의미 단위(발화 단위)로 분리.

⭐ **utt_split**

* 몇 초 이상 침묵하면 새로운 발화로 나눌지 설정.

**version**

* 사용할 모델 버전.

**mip_opt_out**

* Deepgram의 모델 개선 프로그램에 데이터를 사용하지 않도록 설정.

**request_options**

* 요청 자체에 대한 추가 설정(타임아웃 등).

---

### 반환값(Returns)

`MediaTranscribeResponse`

* 일반적으로는 **전사 결과(JSON)** 를 반환한다.
* callback 방식을 사용하는 경우에는 **request_id**를 반환한다.

---

## 네 프로젝트에서 실제로 신경 쓸 옵션

사실 너는 이 정도만 보면 된다.

```python
client.listen.v1.media.transcribe_file(
    request=audio_bytes,
    model="nova-3",
    language="ko-KR",
    filler_words=True,
    punctuate=False,
    smart_format=False,
    utterances=True
)
```

특히 내가 가장 관심 있게 볼 옵션은

* ⭐ `filler_words`
* ⭐ `punctuate`
* ⭐ `smart_format`
* ⭐ `utterances`
* ⭐ `utt_split`

이 다섯 개야.

왜냐하면 발표 평가에서는

* "어...", "음..."을 살려야 하고
* 문장을 AI가 예쁘게 고치면 안 되고
* 어디서 말을 끊었는지도 중요하기 때문이야.

**TL;DR:** 이 함수에는 40개가 넘는 옵션이 있지만, 네 프로젝트에서는 `request`, `model`, `language`, `filler_words`, `punctuate`, `smart_format`, `utterances`, `utt_split` 정도만 알면 충분하다. 나머지는 감정 분석, 요약, 주제 추출, 개인정보 제거 같은 부가 기능이다.
